Workiva/dart_dev
# CP-1811 Add Local Task Integrations

Problem
-------

As a developer I would like to include custom tasks to be used within a single project. To do this I would need to write a minimum set of configs and task classes and register them with dart_dev within the
dev.dart file. It would be nice if there was a local command that could find all tasks within a tool director and allow dart_dev to run the task.

Solution
--------

Provide the Local task within dart_dev that scans the the `tool` directory of the current project and finds files that match the default pattern of `[taskName]_task.[ext]`. Running `pub run dart_dev local
[taskName] [args]` will start a task with a defined executable in the LocalConfig. Consumers can add additional executables that can handle a given `ext` pattern.

Howto Test
----------
- Checkout PR branch
- Run the added local tasks
    - [ ] `pub run dart_dev hello`
- Add a new executable
    -  [ ] Open `dev.dart`
    -  [ ] Add another interpreter or runtime to the list of executables -- *I will demonstrate with ruby but you can use something else if you prefer*
        - Add another value to the `executables` map within the `config.local` object.
        -  <img width="585" alt="screen shot 2016-05-02 at 9 21 25 am" src="https://cloud.githubusercontent.com/assets/11619752/14958289/4df455d8-1047-11e6-9404-1fbb4b9201fb.png">
    - [ ] Create a file that uses the new executable
       - `echo "puts 'hello world" > tool/tasks/rubyTest_task.rb`
    - [ ] Confirm the task is discovered and displayed when no task is given
       - `pub run dart_dev`
       - `rubyTest` should appear in the list of available local tasks
    - [ ] Confirm the task will run
       - `pub run dart_dev rubyTest`
       - You should see the output from the ruby script
       - <img width="471" alt="screen shot 2016-05-02 at 9 30 15 am" src="https://cloud.githubusercontent.com/assets/11619752/14958554/8a184852-1048-11e6-9159-c84914997679.png">

References
----------
- /cc 
    - @evanweible-wf
    - @maxwellpeterson-wf
    - @trentgrover-wf 
    - @georgelesica-wf
